modules/core/src/vaapi.cpp

```diff
@@ -0,0 +1,326 @@
+        status = clGetDeviceIDsFromVA_APIMediaAdapterINTEL(platforms[i], CL_VA_API_DISPLAY_INTEL, display,
+                                                           CL_PREFERRED_DEVICES_FOR_VA_API_INTEL, numDevices, &device, NULL);
+        if (status != CL_SUCCESS)
+            continue;
+
+        // Creating CL-VA media sharing OpenCL context
+
+        cl_context_properties props[] = {
+            CL_CONTEXT_VA_API_DISPLAY_INTEL, (cl_context_properties) display,
+            CL_CONTEXT_INTEROP_USER_SYNC, CL_FALSE, // no explicit sync required
+            0
+        };
+
+        context = clCreateContext(props, numDevices, &device, NULL, NULL, &status);
+        if (status != CL_SUCCESS)
+        {
+            clReleaseDevice(device);
+        }
+        else
+        {
+            found = i;
+            break;
+        }
+    }
+
+    if (found < 0)
+        CV_Error(cv::Error::OpenCLInitError, "OpenCL: Can't create context for VA-API interop");
+
+    Context& ctx = Context::getDefault(false);
+    initializeContextFromHandle(ctx, platforms[found], context, device);
+    return ctx;
+#endif
+}
+
+#if defined(HAVE_VAAPI) && defined(HAVE_OPENCL)
+static bool ocl_convert_nv12_to_bgr(cl_mem clImageY, cl_mem clImageUV, cl_mem clBuffer, int step, int cols, int rows)
+{
+    ocl::Kernel k;
+    k.create("YUV2BGR_NV12_8u", cv::ocl::core::cvtclr_dx_oclsrc, "");
+    if (k.empty())
+        return false;
+
+    k.args(clImageY, clImageUV, clBuffer, step, cols, rows);
+
+    size_t globalsize[] = { cols, rows };
+    return k.run(2, globalsize, 0, false);
+}
+
+static bool ocl_convert_bgr_to_nv12(cl_mem clBuffer, int step, int cols, int rows, cl_mem clImageY, cl_mem clImageUV)
+{
+    ocl::Kernel k;
+    k.create("BGR2YUV_NV12_8u", cv::ocl::core::cvtclr_dx_oclsrc, "");
+    if (k.empty())
+        return false;
+
+    k.args(clBuffer, step, cols, rows, clImageY, clImageUV);
+
+    size_t globalsize[] = { cols, rows };
+    return k.run(2, globalsize, 0, false);
+}
+#endif // HAVE_VAAPI && HAVE_OPENCL
+
+} // namespace cv::vaapi::ocl
+
+void convertToVASurface(InputArray src, VASurfaceID surface, Size size)
+{
+    (void)src; (void)surface; (void)size;
+#if !defined(HAVE_VAAPI)
+    NO_VAAPI_SUPPORT_ERROR;
+#elif !defined(HAVE_OPENCL)
+    NO_OPENCL_SUPPORT_ERROR;
+#else
+    const int stype = CV_8UC4;
+
+    int srcType = src.type();
+    CV_Assert(srcType == stype);
+
+    Size srcSize = src.size();
+    CV_Assert(srcSize.width == size.width && srcSize.height == size.height);
+
+    UMat u = src.getUMat();
+
+    // TODO Add support for roi
+    CV_Assert(u.offset == 0);
+    CV_Assert(u.isContinuous());
+
+    cl_mem clBuffer = (cl_mem)u.handle(ACCESS_WRITE);
+
+    using namespace cv::ocl;
+    Context& ctx = Context::getDefault();
+    cl_context context = (cl_context)ctx.ptr();
+
+    cl_int status = 0;
+
+    cl_mem clImageY = clCreateFromVA_APIMediaSurfaceINTEL(context, CL_MEM_WRITE_ONLY, &surface, 0, &status);
```

Comment:
To use interop, you should call initializeContextFromVA() function prior to call convertFrom/To, this is noted in short feature documentation (vaapi.hpp). initialize() iterates over CL platforms and tries to find such that
- extension pointers (clGetExtensionFunctionAddressForPlatform) all are not NULL &&
- has compatible device (CL_PREFERRED_DEVICES_FOR_VA_API_INTEL) &&
- CL context created (CL_CONTEXT_VA_API_DISPLAY_INTEL).
In case of absence of such platform, CV_Error called.
Perhaps, in convertFrom/To functions we should check if initialization is done successfully.